add documentation for running wasm workloads (#3161)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/topics/wasi.md

# WebAssembly / WASI Workloads

## Overview

> CAPZ enables you to create WebAssembly (Wasm) / WASI pod workloads targeting either [Deislabs Slight](https://github.com/deislabs/spiderlightning) or [Fermyon Spin](https://github.com/fermyon/spin) frameworks for building and running fast, secure microservices with Wasm on Kubernetes (v1.23.16+, v1.24.10+, v1.25.6+, v1.26.1+, and newer Kubernetes versions).

Reviewer: ```suggestion
CAPZ enables you to create WebAssembly (Wasm) / WASI pod workloads targeting either [Deislabs Slight](https://github.com/deislabs/spiderlightning) or [Fermyon Spin](https://github.com/fermyon/spin) frameworks for building and running fast, secure microservices on Kubernetes (v1.23.16+, v1.24.10+, v1.25.6+, v1.26.1+, and newer Kubernetes versions).
```

Both of the runtimes (slight and spin) for running Wasm workloads use [Wasmtime](https://wasmtime.dev) embedded in containerd shims via the [deislabs/containerd-wasm-shims](https://github.com/deislabs/containerd-wasm-shims) project which is built upon [containerd/runwasi](https://github.com/containerd/runwasi). These containerd shims enable Kubernetes to run Wasm workloads without needing to embed the Wasm runtime in each OCI image. 

## Slight (SpiderLightning)

> Slight (or Spiderlightning) is an open source wasmtime-based runtime that provide cloud capabilities to Wasm microservices. These capabilities include key/value, pub/sub, and much more.

Reviewer: ```suggestion
Slight (or Spiderlightning) is an open source wasmtime-based runtime that provides cloud capabilities to Wasm microservices. These capabilities include key/value, pub/sub, and much more.
```

> "SpiderLightning defines a set of WebAssembly Interface Types (i.e., WIT) files that abstract distributed application capabilities, such as state management, pub/sub, event driven programming, and more."

Reviewer 2: May need to leave a more descriptive way to introduce `slight`. 

"slight is a open source wasmtime-based runtime that provide cloud capabilities to Wasm microservices. These capabilites include keyvalue, pub/sub, and much more. "

Author: Great suggestion.

## Fermyon Spin
"Spin is an open source framework for building and running fast, secure, and composable cloud microservices with WebAssembly. It aims to be the easiest way to get started with WebAssembly microservices, and takes advantage of the latest developments in the WebAssembly component model and Wasmtime runtime."

### Applying the Wasm Runtime Classes

> By default, CAPZ virtual machines include containerd shims to run both `slight` and `spin` workloads. To inform Kubernetes about the ability to run these workloads on CAPZ nodes, you will need to apply a runtime class for each runtime.

Reviewer 3: ```suggestion
By default, CAPZ reference virtual machine images include containerd shims to run both `slight` and `spin` workloads. To inform Kubernetes about the ability to run these workloads on CAPZ nodes, you will need to apply a runtime class for each runtime.
```

Reviewer 3: Actually, is there a minimum image-builder / k8s version for which our images have this? My first attempt failed with

> Failed to create pod sandbox: rpc error: code = Unknown desc = failed to get sandbox runtime: no runtime for "spin" is configured

I'm using k8s v1.26.0

Reviewer 3: looks like https://github.com/kubernetes-sigs/image-builder/pull/1037 went in Dec 22nd and v1.26.0 was published Dec 12th so I might need a newer image (I wouldn't expect a user to know how to find this out easily so worth documenting IMO)

Author: I believe I call out the K8s versions in the overview section. Is that sufficient?

Author: Per your error, that is expected given there is no runtime configured for spin in the containerd config.

Reviewer: I think only the Azure ref images for v1.26.1 and later have the WASM shims included.

Author: @mboersma why would they not have been included in say v1.23.16 and others that were published at the same time as 1.26.1?

Reviewer: @devigned, sorry, I misspoke. I was thinking for Cecile that she needs v1.26.1, but all the recent set of patches (v1.23.16 and friends) have the Wasm shims.

Reviewer 3: > I believe I call out the K8s versions in the overview section. Is that sufficient?

Ah I somehow missed that (or rather read it then forgot about it by the time I got the next step 😬 . Should be good enough!

> By default, CAPZ virtual machines include containerd shims to run both `slight` and `spin` workloads. To inform Kubernetes about the ability to run these workloads on CAPZ nodes, you will need to apply a runtime class for each runtime.

Reviewer 3: > you will need to apply a runtime class for each runtime

I'm assuming we're talking about applying the manifest on the workload cluster here? could be more explicit

```yaml
---
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: "wasmtime-slight-v1"
handler: "slight"
---
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: "wasmtime-spin-v1"
handler: "spin"
```

> The preceding runtime classes will register a runtime class for slight and spin, which will direct containerd to use the spin or slight shim when a pod workload is scheduled onto a cluster node.

Reviewer: ```suggestion
The preceding YAML document will register a runtime class for slight and spin, which will direct containerd to use the spin or slight shim when a pod workload is scheduled onto a cluster node.
```

### Running an Example Spin Workload
With the runtime classes registered, we can now schedule Wasm workloads on our nodes.

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wasm-spin
spec:
  replicas: 3
  selector:
    matchLabels:
      app: wasm-spin
  template:
    metadata:
      labels:
        app: wasm-spin
    spec:
      runtimeClassName: wasmtime-spin-v1
      containers:
        - name: spin-hello
          image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
          command: ["/"]
          resources:
            requests:
              cpu: 10m
              memory: 10Mi
            limits:
              cpu: 500m
              memory: 128Mi
---
apiVersion: v1
kind: Service
metadata:
  name: wasm-spin
spec:
  type: LoadBalancer
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  selector:
    app: wasm-spin
```

> The preceding deployment and service will create a load balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.

Reviewer: ```suggestion
The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.
```

> The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.

Author: I added a section to describe how to curl the services and how to get started building `slight` and `spin` microservices.

Thank you again for pointing out that the URI was not obvious.

> The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.

Reviewer: Thanks for adding the new section! I followed the instructions robotically and succeeded:

```
Hello world from Spin!
```

### Constraining Scheduling of Wasm Workloads
You may have a cluster where not all nodes are able to run Wasm workloads. In this case, you would want to constrain the nodes that are able to have Wasm workloads scheduled. 

If you would like to constrain the nodes that will run the Wasm workloads, you can apply a node label selector to the runtime classes, and apply node labels to the cluster nodes you'd like to run the workloads.

```yaml
---
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: "wasmtime-slight-v1"
handler: "slight"
scheduling:
  nodeSelector:
    "cluster.x-k8s.io/wasmtime-slight-v1": "true"
---
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: "wasmtime-spin-v1"
handler: "spin"
scheduling:
  nodeSelector:
    "cluster.x-k8s.io/wasmtime-spin-v1": "true"
```

> In the preceding yaml, note the nodeSelector and the label. The Kubernetes scheduler will select nodes with the `cluster.x-k8s.io/wasmtime-slight-v1: "true"` or the `cluster.x-k8s.io/wasmtime-spin-v1: "true"` to determine where to schedule Wasm workloads.

Reviewer: ```suggestion
In the preceding YAML, note the nodeSelector and the label. The Kubernetes scheduler will select nodes with the `cluster.x-k8s.io/wasmtime-slight-v1: "true"` or the `cluster.x-k8s.io/wasmtime-spin-v1: "true"` to determine where to schedule Wasm workloads.
```

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wasm-spin
spec:
  replicas: 3
  selector:
    matchLabels:
      app: wasm-spin
  template:
    metadata:
      labels:
        app: wasm-spin
    spec:
      runtimeClassName: wasmtime-spin-v1
      containers:
        - name: spin-hello
          image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
          command: ["/"]
          resources:
            requests:
              cpu: 10m
              memory: 10Mi
            limits:
              cpu: 500m
              memory: 128Mi
---
apiVersion: v1
kind: Service
metadata:
  name: wasm-spin
spec:
  type: LoadBalancer
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  selector:
    app: wasm-spin
```

The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.

### A Running Spin Microservice

> With the service and the deployment applied, you should now have a running Spin microservice running in your workload cluster. If you run the following command against the workload cluster, you can find the IP for the `wasm-spin` service.

Reviewer: ```suggestion
With the service and the deployment applied, you should now have a Spin microservice running in your workload cluster. If you run the following command against the workload cluster, you can find the IP for the `wasm-spin` service.
```

```shell
kubectl get services -w
NAME         TYPE           CLUSTER-IP      EXTERNAL-IP     PORT(S)        AGE
kubernetes   ClusterIP      10.96.0.1       <none>          443/TCP        14m
wasm-spin    LoadBalancer   10.105.51.137   20.121.244.48   80:30197/TCP   3m8s
```

In the preceding output, we can see the `wasm-spin` service with an external IP of `20.121.244.48`. Your external IP will be a different IP address, but that is expected.

Next, let's curl the service and get a response from our Wasm microservice. You will need to replace the placeholder IP address with the external IP address from the preceding output.

```shell
curl http://20.121.244.48/hello
Hello world from Spin!
```

In the preceding output, we see the HTTP response from our Spin microservice, "Hello world from Spin!".

### Building a Spin or Slight Application

> At this point, you might be asking how do I build my own Wasm microservice. Here are a couple pointers to help you get started.

Reviewer: ```suggestion
At this point, you might be asking "How do I build my own Wasm microservice?" Here are a couple pointers to help you get started.
```
